### PR TITLE
(PUP-11373) Always lock the environment before its type collection

### PR DESCRIPTION
**Pull request by joshcooper**
Puppetserver in MT mode can deadlock if one thread calls into the loaders,
which locks the environment, and then tries to lock the type_collection:

    - waiting to lock <3f27e688> (a org.jruby.gen.RubyObject0) owned by "qtp346358872-131" t@131
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/type_collection.rb:188
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/type_collection.rb:152
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/type_collection.rb:152
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/pops/loader/runtime3_type_loader.rb:79
    ...
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/concurrent/lock.rb:10
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/pops/loader/loader.rb:153

While another thread locks the type_collection and tries to lock the environment:

    - waiting to lock <bd0f663> (a org.jruby.gen.RubyObject0) owned by "qtp346358872-67" t@76
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/node/environment.rb:248
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/parser/type_loader.rb:131
    ...
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/type_collection.rb:203
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/concurrent/lock.rb:10
    /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/type_collection.rb:188

This commit orders locking so we always acquire the environment lock before its
type collection lock.

### CODE REVIEW

**Comment by joshcooper** [comment]
This targets `main` because the loaders only started locking the environment in https://github.com/puppetlabs/puppet/commit/ec319d778150eade0b6128c2ddd11cbf3926183b first released in 7.6.0

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Review by mwaggett — COMMENTED** [review]
This looks reasonable to me, though someone more involved with the multithreaded work may want to take a look.

Is there a test that could demonstrate the bad behavior and then prove that this fixes it? Or is a manual test more appropriate?

**Comment by joshcooper** [comment]
@mwaggett thanks for looking! The OP that filed the ticket reported that it was [working after 5+ hours](https://tickets.puppetlabs.com/browse/PUP-11373?focusedCommentId=838391&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-838391) It's possible to add a test, but I think it'd take me a bit to figure that out. There's at least one other remaining MT issue (the "attempt to redefine entity" on puppetserver) so I'm inclined to merge this and "move the needle forward". Thoughts?

File: lib/puppet/resource/type_collection.rb [hunk 1]
```diff
@@ -185,26 +186,29 @@ def version
...
-      result = send(type, fqname)
-      unless result
-        if @notfound[ fqname ] && Puppet[ :ignoremissingtypes ]
-          # do not try to autoload if we already tried and it wasn't conclusive
-          # as this is a time consuming operation. Warn the user.
-          # Check first if debugging is on since the call to debug_once is expensive
-          if Puppet[:debug]
-            debug_once _("Not attempting to load %{type} %{fqname} as this object was missing during a prior compilation") % { type: type, fqname: fqname }
+    # always lock the environment before locking the type collection
+    @environment.lock.synchronize do
```
Comment (justinstoller) [inline lib/puppet/resource/type_collection.rb:190]: Is there a point to using the type collection lock at all now?

**Review by justinstoller — APPROVED** [review]
If you have someone successfully using this I'm happy to have any progress made on this since we were at our wits end just trying to get a reproducer. Though I would love to see more about how to reproduce as we have some scripts that we've used to try and catch these before here that could use updating: https://github.com/puppetlabs/puppetserver/tree/6.x/ext/thread_test

**Comment by mwaggett** [comment]
seconding Justin - fine with me to get this in if we've got a user successfully running with this, but any updates to the testing would be appreciated (though also fine to do that in a separate PR, if at all).

File: lib/puppet/resource/type_collection.rb (hunk 1 above)
Comment (joshcooper) [reply lib/puppet/resource/type_collection.rb:190]: It's probably not needed, but I don't think it's hurting anything given the lock ordering. I'll keep this in mind for future.

**Comment by joshcooper** [comment]
Thanks @justinstoller and @mwaggett! I added a ticket to update the test as SERVER-3155. I'll take a look at that a bit later.

**Pull request merged**